internal/factsengine/gatherers/hostsfile.go

```diff
@@ -0,0 +1,138 @@
+package gatherers
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/agent/internal/factsengine/entities"
+)
+
+const (
+	HostsFileFactKey    = "hosts"
+	HostsFilePath       = "/etc/hosts"
+	ipMatchGroup        = "ip"
+	hostnamesMatchGroup = "hostnames"
+)
+
+var (
+	parsingRegexp      = `(?m)(?P<` + ipMatchGroup + `>\S+)\s+(?P<` + hostnamesMatchGroup + `>.+)`
+	HostsEntryCompiled = regexp.MustCompile(parsingRegexp)
+	//HostsEntryCompiled = regexp.MustCompile(`(?m)(?P<ip>\S+)\s+(?P<hostnames>.+)`)
+)
+
+// nolint:gochecknoglobals
+var (
+	HostsFileError = entities.FactGatheringError{
+		Type:    "hosts-file-error",
+		Message: "error reading /etc/hosts file",
+	}
+
+	HostsFileDecodingError = entities.FactGatheringError{
+		Type:    "hosts-file-decoding-error",
+		Message: "error decoding /etc/hosts file",
+	}
+
+	HostsEntryNotFoundError = entities.FactGatheringError{
+		Type:    "hosts-file-value-not-found",
+		Message: "requested field value not found in /etc/hosts file",
+	}
+)
+
+type HostsFileGatherer struct {
+	hostsFile string
+}
+
+func NewDefaultHostsFileGatherer() *HostsFileGatherer {
+	return NewHostsFileGatherer(HostsFilePath)
+}
+
+func NewHostsFileGatherer(hostsFile string) *HostsFileGatherer {
+	return &HostsFileGatherer{hostsFile: hostsFile}
+}
+
+func (s *HostsFileGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := []entities.Fact{}
+	log.Infof("Starting /etc/hosts file facts gathering process")
+
+	hostsFile, err := readHostsFileByLines(s.hostsFile)
+	if err != nil {
+		return nil, HostsFileError.Wrap(err.Error())
+	}
+
+	hostsFileMap, err := hostsFileToMap(hostsFile)
+	if err != nil {
+		return nil, HostsFileDecodingError.Wrap(err.Error())
+	}
+
+	for _, factReq := range factsRequests {
+		var fact entities.Fact
+		var found bool
+
+		for hostname, ip := range hostsFileMap {
+			if hostname == factReq.Argument {
+				fact = entities.NewFactGatheredWithRequest(factReq, ip)
+				facts = append(facts, fact)
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, errors.New(HostsEntryNotFoundError.Error())
```

Comment:
You must continue looping here even though you have 1 failure. There might be other facts for other checks.
You could have something like:
```
gatheringError := HostsEntryNotFoundError.Wrap(factReq.Argument)
log.Error(gatheringError)
fact = entities.NewFactGatheredWithError(factReq, gatheringError)
```